web/src/location.ts

```diff
@@ -0,0 +1,81 @@
+import { advanceLocationSimulation, locationSimulationFromRoute } from "ferrostar";
+
+export class SimulatedLocationProvider {
+  private simulationState = null;
+
+  lastLocation = null;
+  lastHeading = null;
+  warpFactor = 1;
+
+  updateCallback: () => void = () => {};
+
+  setSimulatedRoute(route: any) {
+    this.simulationState = locationSimulationFromRoute(route, 10.0);
+    this.startSimulation();
+  }
+
+  async startSimulation() {
+    while (true) {
+      await new Promise((resolve) => setTimeout(resolve, (1 / this.warpFactor) * 1000));
+      const initialState = this.simulationState;
+      const updatedState = advanceLocationSimulation(initialState);
+
+      if (initialState === updatedState) {
+        return;
+      }
+
+      this.simulationState = updatedState;
+      this.lastLocation = updatedState.current_location;
+
+      // Since Lit cannot detect changes inside objects, here we use a callback to trigger a re-render
+      // This is a minimal approach if we don't want to use a state management library like MobX, but might not be the ideal solution
+      if (this.updateCallback) {
+        this.updateCallback();
+      }
+    }
+  }
+
+  stopSimulation() {
+    this.simulationState = null;
+  }
+}
+
+export class BrowserLocationProvider {
+  private geolocationWatchId: number | null = null;
+  lastLocation: any = null;
+  lastHeading = null;
+  warpFactor = 1;
+
+  updateCallback: () => void = () => {};
+
+  requestPermission() {
+    navigator.permissions.query({ name: "geolocation" });
```

Comment:
This will pop up a permission request window on the browser. However, I just found that `navigator.geolocation.watchPosition` can trigger the prompt as well. I will look into if we can remove this function.

Reply:
Yeah, as far as I can tell, calls like `watchPosition` will result in browser prompts when needed. I'd say we can remove this method, unless the point is to update some state which is observed elsewhere (ex: to show a "location unavailable" indicator on the page, but that's not an immediate requirement.)